rework evo stop conditions 🛑

SWeav02 opened the pull request:
This PR primarily aims to add a stop condition to the evolutionary search workflow when the search reaches the expected structure. It also aims to implement some fixes from #726 to the packing factor variable in the random structure generator.

The main steps for the stop conditions are:

- [x] Build out [StopCondition](https://github.com/jacksund/simmate/blob/91af10b06c5bc37f045aa4d3b15d5453377affa0/src/simmate/apps/evolution/stop_conditions/stop_conditions.py) class
- [x] Move  [search.check_stop_condition](https://github.com/jacksund/simmate/blob/91af10b06c5bc37f045aa4d3b15d5453377affa0/src/simmate/apps/evolution/models/fixed_composition.py#L116-L174) logic to BasicStopCondition class
- [x] Add StopCondition for finding expected structure

The main steps for the packing factor are:

- [x] test series of packing factors on small structures
- [x] select most successful

SWeav02 committed:
small fixes to evo-search summary and staged workflow

SWeav02 committed:
switch packing factor

SWeav02 committed:
Update StopCondition class

SWeav02 committed:
Add stop conditions to FixedCompositionSearch

SWeav02 committed:
Merge branch 'jacksund:main' into minor_evo_fixes

SWeav02 committed:
update docs and tests

SWeav02 commented:
@jacksund This PR is ready for your review. I formatted with Black on my end so I'm assuming there was some update that's causing the tests to fail.

I mostly followed what we discussed in #726. I ended up adding another column to the FixedCompositionSearch table for stop conditions. I also set it up so that the input is defined like `stop_conditions={stop_condition1: kwargs1, stop_condition2: kwargs2, ....}` but at the moment our two conditions (BasicStopConditions and ExpectedStructure) don't need any kwargs.

Otherwise, I just made a couple of tweeks to fix problems with the search summary.

jacksund commented:
yep, there's a new black release that I'll fix in things in a separate PR and rebasing this pr will get the ci working again

jacksund commented:
> (BasicStopConditions and ExpectedStructure) don't need any kwargs.

So you have it where the yaml will look like this:

``` yaml
workflow_name: structure-prediction.toolkit.fixed-composition
composition: Ca8N4F8
expected_structure: example.cif
max_structures: 15000
stop_conditions:
    ExpectedStructure: {}
    BasicStopConditions: {}
```

right? That looks wacky to me. I expected this:

``` yaml
workflow_name: structure-prediction.toolkit.fixed-composition
composition: Ca8N4F8
stop_conditions:
    ExpectedStructure:
        expected_structure: example.cif
    BasicStopConditions:
        max_structures: 15000
```

The way you have things set up, it might make sense have `stop_conditions` be a list, but this makes it difficult for anyone to add custom settings & users would need to update settings in two places to get something to work. What do you think:

``` yaml
workflow_name: structure-prediction.toolkit.fixed-composition
composition: Ca8N4F8
expected_structure: example.cif
max_structures: 15000
stop_conditions:
    - ExpectedStructure
    - BasicStopConditions
```

jacksund commented on src/simmate/apps/evolution/stop_conditions/stop_conditions.py:
```diff
@@ -21,35 +27,114 @@ def check(self, search):
+            return False
+        expected_structure = Structure.from_dynamic(expected_structure_dynamic)
+        # get completed individuals
+        individuals = search.individuals_completed.order_by("finished_at").all()
```
yeah this is a bottleneck and wasn't an issue when I ran these checks manually. One workaround is caching the last check (as a class attribute, not in the database), and then doing `.individuals_completed.filter(finished_at__gte=last_check_timestamp)`  to only check the new ones

jacksund commented on src/simmate/apps/evolution/stop_conditions/stop_conditions.py:
```diff
@@ -21,35 +27,114 @@ def check(self, search):
+                break
+
+        if not is_match:
+            logging.info("Search has not found groundstate yet")
```
update the logging to be "expected structure", not "groundstate". My original print statements were groundstate but your ExpectedStructure class is more flexible

jacksund commented on src/simmate/engine/staged_workflow.py:
```diff
@@ -95,30 +100,38 @@ def run_config(
         # save final result
-        final_result = (
-            dict(
+        if error is None:
```
is this related to the stop conditions at all? I'm not sure what's going on here. But since refactoring staged workflows is on my to-do list, I'm included to say whatever and just pass this through & clean up later haha

SWeav02 replied:
It's not lol. I forgot about some of the smaller things I did before I started working on the stop conditions. This was just to avoid an annoying error readout that I get throughout the search.

jacksund commented on src/simmate/toolkit/creators/lattice.py:
```diff
@@ -179,6 +179,7 @@ def __init__(
         vector_gen_options=dict(),
         angle_generation_method=UniformlyDistributedVectors,
         angle_gen_options=dict(min_value=60, max_value=120),
+        packing_factor: int = 0.74,
```
I'm biased toward **kwargs when the defaults are the same, but this works too

SWeav02 replied:
I'm happy to switch over to **kwargs. This is just a leftover from some of my testing that I figured I'd leave for convenience.

SWeav02 commented:
This is indeed what it currently looks like:
```yaml
workflow_name: structure-prediction.toolkit.fixed-composition
composition: Ca8N4F8
expected_structure: example.cif
max_structures: 15000
stop_conditions:
    ExpectedStructure: {}
    BasicStopConditions: {}
```
And actually I had also envisioned something more like this:
```yaml
workflow_name: structure-prediction.toolkit.fixed-composition
composition: Ca8N4F8
stop_conditions:
    ExpectedStructure:
        expected_structure: example.cif
    BasicStopConditions:
        max_structures: 15000
```
It ended up like the former because I didn't want to stray too far from the scheme we currently have. I originally set up the `ExpectedStructure` to work exactly like you noted in the latter, but I realized that an `expected_structure` column already existed in the `FixedCompositionSearch` table, so it was easier to add it as a separate parameter. I had the same sort of thought for the `BasicStopConditions` where I didn't want to rework the way the max_structures, min_structures, etc. parameters worked, so I left them as they were.

I wouldn't mind reworking things to behave more like the second option, but I think it would require adding a step where we loop over the stop conditions and save any parameters to the database. That might be a little ugly since not all parameters  will necessarily have columns in the database, so we'd have to sort them somehow. Thoughts?

I still prefer either of these options to using a list for the same reasons you mentioned related to making it difficult to customize.

jacksund commented:
You made `stop_conditions` a JSONField and we will be storing all parameters within the json. Then we can delete the `max_structures`, `expected_structure`, etc. columns because they are now in the JSON field

SWeav02 commented:
> You made `stop_conditions` a JSONField and we will be storing all parameters within the json. Then we can delete the `max_structures`, `expected_structure`, etc. columns because they are now in the JSON field

This is obvious in retrospect 😆

I'll change it tomorrow

SWeav02 committed:
update stop_conditions parameters

SWeav02 committed:
add last_check_timestamp

SWeav02 committed:
Update chemical_system and variable_nsites_composition

SWeav02 committed:
black and isort

SWeav02 commented:
I forgot to move this back over to a draft before adding more commits :disappointed: 

I'm going to adjust the parameters and evolution docs to match with the new StopCondition input scheme and then I'll reopen the PR for review.

SWeav02 committed:
update docs and tests

jacksund committed:
Merge branch 'main' into minor_evo_fixes

jacksund commented on src/simmate/apps/evolution/workflows/variable_nsites_composition.py:
which error? I can't see the code your comment is attached to. Maybe you mean the `started_at` issue?

jacksund replied:
got it. I think I know how to fix this and will take a stab at it

SWeav02 commented on src/simmate/apps/evolution/workflows/chemical_system.py:
```diff
@@ -155,7 +155,7 @@ def run_config(
         # ---------------------------------------------------------------------
         # Submitting known structures
         # ---------------------------------------------------------------------
-
```
I added these to avoid an error when no singleshot sources are provided for the chemical_system search. In the original workflow these variables are only defined behind if statements and if they don't get triggered this line throws an error later:
https://github.com/jacksund/simmate/blob/9819548182e7ee829ff4b41d946446855c590c83/src/simmate/apps/evolution/workflows/chemical_system.py#L211
So the lists are just placeholders to avoid the error.

jacksund committed:
condense migrations
